Billy

GrandeJu opened the pull request:
have a look

Commit:
数据库设计文档添加

Commit:
Fix: CRLF

Commit:
database

rtdarwin commented:
达不到数据库设计初稿的标准

### 格式：
1. 中英文之间缺少空格
2. 行尾错用 CRLF，改为 LF
3. 从 4.1 节开始，表格中空格数量混乱

### 内容：
1. 第三章节贴图改用 PowerDesigner 的 CDM 图，并将 CDM 原件上传至本库
2. 缺少外码约束，缺少触发器等
3. 缺少 DDL ，无法根据文档建立数据库。参考 MySQL 的手册建立数据库，再把建立数据库所用 SQL 语句放入一 .sql 结尾的文件内上传本库（代码库建立之前暂存至此）
4. 删除复制过来的第 6、7 章。优化决策待 @c-rainstorm 添加

### 关系模式/属性 名：
1. 收货人表：consignee
2. 购物车表：shopping_cart
3. 分类表：category
4. 用户表::头像：avatar
5. 事务表::提交人：committer
6. 事务表::处理备注：annotation
7. 订单表::备注：annotation
8. 订单状态用 status，事务状态用 transaction_status？不一致
9. 商品评价表名 goods_evaluate，订单表::评价编号名 commit_id，店铺表::评价人数用 evaluate_number？不一致
10. 评价分值名 commment_score，店铺::用户评价综合名 credit_rating？不一致

### 修改以上格式与结构的不一致之后，数据库关系模式的一致性再由 @c-rainstorm  review。

@ytqqwer @billyblondking

Commit:
数据库设计完善

rtdarwin committed:
CRLF -> LF

rtdarwin committed:
fix typo

rtdarwin commented:
文档的格式和结构已完成，关系模式方面：
1. 未完全实现需求文档要求的约束，且与需求文档有不一致的地方
2. 尚缺少保证数据库数据完整性的触发器等措施

@c-rainstorm

c-rainstorm commented:
# 数据库文档修改建议 1

<!-- TOC -->

- [数据库文档修改建议](#%E6%95%B0%E6%8D%AE%E5%BA%93%E6%96%87%E6%A1%A3%E4%BF%AE%E6%94%B9%E5%BB%BA%E8%AE%AE)
    - [『数据库设计.md』修改建议](#%E3%80%8E%E6%95%B0%E6%8D%AE%E5%BA%93%E8%AE%BE%E8%AE%A1md%E3%80%8F%E4%BF%AE%E6%94%B9%E5%BB%BA%E8%AE%AE)
    - [『create_database.sql』修改建议](#%E3%80%8Ecreate_databasesql%E3%80%8F%E4%BF%AE%E6%94%B9%E5%BB%BA%E8%AE%AE)

<!-- /TOC -->

---

## 『数据库设计.md』修改建议

1. 『1.2 文档范围』中，在列表项之前添加空行，不然无法正确生成有序列表。
1. 『2. 数据库系统说明』中，运行目标系统改为 Windows 10 + Debian 8.6 。
1. 『4. 物理设计』中，注册号不是由管理员发放，而是由用户申请开店时直接需要的填写的项目。
1.  **因 MySQL 中不支持用户自定义域，所以所有『域』统一改为『枚举』。（表格中数据类型改为枚举，说明中填写枚举值）**
1. 『4.1 用户表』中，删除『店铺编号』。店铺表中已经包含了 user_id 这一外键。
1. 『4.3 事务表』中，『事务类型域』中删去『意见反馈』；『事物状态域』改为『未处理、已通过、已拒绝』。
1. 『4.4 店铺表』中，『平均评分』改为『总评分』。
1. 『4.5 商品订单表』中，『订单状态』改为『待付款、待发货、待收货、待评价、已完成、已取消』。此修改造成的与需求文档的不一致见『其他』一节所述。
1. 『4.6 订单中商品表』中，商品编号不唯一，不可为主键；解决方案：
    1. 可添加一个 id 属性做主键（推荐）；
    2. 商品编号 + 订单编号作联合主键；
1. 『4.8 商品表』中『预览图』改为『预览图数量』。图片名称使用统一格式『商品编号-图片编号.jpg』。例如 1000-3.jpg 表示商品 1000 的第 3 张图。『打折截止时间』默认为 `2000-01-01 00:00:00`，『打折比例』默认为 `1` 。
    - 这种方式其实也不是很好，因为添加商品是商品编号是未知的，需要将商品信息插入到表中以后再将商品编号 `select` 出来，稍有些麻烦。如果有更好地方式的话可以作相应修改！
1. 『4.9 收藏店铺表』 主键设置不合理，同 4.6。
1. 『4.10 收藏商品表』主键设置不合理，同 4.6。
1. 『4.11 购物车表』主键设置不合理，同 4.6。
1. 部分表未考虑假删除，需要添加状态属性 `is_valid`，`true` 表示有效，`false` 表示无效。

//TODO： 为商品属性添加适当的表。稍后我会再提一个 commit。专门描述这些表。

## 『create_database.sql』修改建议

除了刚才说到的那些以外，还有几个在 SQL 中需要修改的地方：

1. 以需求文档中数据约束为准，添加 `check` 约束。`password` 属性除外，因为密码存储时是已进行过加密的，一定不符合约束条件。
1.  除一些需要固定长度的字段使用 `char` 类型以外，其他字符串类型的字段全以 `varchar` 类型定义。
1. 在所有表的常用的搜索项上添加索引。比如 『用户表』中经常会使用手机号来查询用户信息，所以可以将手机号作为索引。
1.  `on update restrict;` 的含义是当父表更新时，直接拒绝操作。即不允许父表更新，` on delete restrict` 同理，所以外键约束都写成这种形式是不合理的，需要重新定义。
1. 触发器部分
    1. 可以用来做级联假删除。比如某一商品删除后，使用触发器将收藏夹中相应记录也标记为删除状态。
    1. 商品评分设置后直接将其添加到店铺总评分中，并评价人数 +1 
    1. ...

## 其他

1. 前端（对用户） / 后端（对前端） / 数据库（对后端） 三方都需要对数据进行校检。为了效率考虑的话，可以把后端和数据库其中一方的校检去掉。但几乎完全不能信任和依赖前端的 JavaScript 检查
    1. 用户方运行的 JavaScript 可以被禁用和被用户自行修改
    1. HTTP message 的伪造极其简单
1. 数据库设计过程中会发现需求文档的错误，并造成与需求文档的不一致，数据库初稿完成后要再修正需求文档。 @rtdarwin

---

@rtdarwin 的修改建议已合入本建议。

@billyblondking @ytqqwer 
- 大致看一下修改建议。如果有问题，我们再讨论；如果没有问题抓紧时间修订文档！我们时间不多了！

rtdarwin commented:
### 『数据库设计文档.md』修改建议中：

1. 第 2 点：运行目标系统建议为：`Windows 10 + Debian 8.6`
1. 第 5、7 点：『订单状态』与『事务状态』 与需求文档不一致，以哪种为准？
1. 第 9 点：需要对预览图数量和大小进行约束。如果为效率考虑的话，所有的图片也不能放入同一文件夹下

@c-rainstorm 

1. 前端（对用户） / 后端（对前端） / 数据库（对后端） 三方都需要对数据进行校检。为了效率考虑的话，可以把后端和数据库其中一方的校检去掉。但几乎完全不能信任和依赖前端的 JavaScript 检查
    1. 用户方运行的 JavaScript 可以被禁用和被用户自行修改
    1. HTTP message 的伪造极其简单
1. 除修改建议提到的外，还有：
    1. 店铺表中`店铺名称` 与 `店铺地址` 是否可为空的不一致。个人倾向于使用数据库设计文档中的不可为空
1. 补充如下：

### 关系模式中：

1. 用户表包含店铺编号，店铺表包含用户编号 的冗余。
1. 商品表打折截止时间在创建商品时需要数据库自动生成或后端处理，打折比例在添加商品时默认值需要定义？
1. 如果考虑假删除，建议使用 true 作为有效，false 作为无效。比如商品表中商品状态的默认值（现为 false）

@billyblondking  @ytqqwer

c-rainstorm commented:
## 数据库文档修改建议 2

1. 『4.1 用户表』添加用户名属性 `username`，约束参照需求说明文档数据字典,`user_name` 改为 `nickname`。
1. 『4.3 事务表』中 `商家投诉` 改为 `用户投诉`。
1. 『4.5 商品订单表』运单号为 12 位纯数字，所以使用长整型存储。
1. 『4.6 订单中商品』评价分值类型改为 `short Integer`。
1. `char` 改为 `varchar`。
1. 『店铺表』中 `店铺名`、`店铺地址`在申请开店时填写，并且不可为空。
1. 因为 enum 类型不是 SQL 标准的一部，只在 MySQL 中有。如此定义不利于数据库移植，所以将枚举类型改为 `varchar` 类型，长度自行定义。给这些属性添加 `Check` 约束。

Commit:
按照修改建议进行修改（部分）

未完成的部分：

- 增添触发器
- 建立索引
- 外键约束的修改

c-rainstorm commented:
## 数据库修改建议 3

1. 关于表名
    - collect_goods --> favorite_goods
    - collect_shop --> favorite_shops
    - goods_order --> order
    - to_purchase --> shopping_cart
1. 按照需求文档上**数据格式**要求，对相应数据加 `check` 约束。 
1. 对需求文档上没有格式要求的属性添加必要的 `check` 约束。比如 `库存量 >= 0` 等。
1. `category` 
    - `char(10) --> varchar(10)`。
    - ... **所有**不是固定长度的字符串类型**统一**改为 `varchar()`。
1. 表示状态信息（是否已删除）的属性，**一律**改为 `is_valid`。
1. `shop_id --> registration_id`。
1. `goods`
    - `preview`。 `varchar(20) --> tinyint`。加 `check` 约束，0 - 6。
1. `goods_attribute`
    - `attribute_value`。`varchar(10) --> varchar(40)`。
1. `goods_in_order`
    - `centent --> comment`。
1. `shopping_cart` 中需保存商品属性，修改建议：
    1. 添加 `attribute_id`（推荐）。
    1. `goods_id --> attribute_id`。
1. `user` 
    - `shop_id char(15) --> shop_opening bool default false`。
    - `username` 加 `uinque` 约束。
1. **SQL 语句与数据库文档说明不一致（部分属性名和数据类型），从上到下一个表一个表的核对！**

- 以后修订就直接修订 SQL 语句，不用再折腾 CDM 了。

@rtdarwin 补充？
@ytqqwer @billyblondking 修订！可以对修改建议提出质疑，但请尽量不要忽略我的建议，**每一条**！

c-rainstorm commented:
- 部分表的属性顺序可以改一下。把重要的属性尽量放在前面，并使相邻属性尽量满足一定的逻辑关系。

Commit:
数据库再修改

rtdarwin committed:
数据库再再修改

rtdarwin commented:
## 数据库设计文档 进度报告

已大致完成修改建议 1，修改建议 2，修改建议 3 的业务方面的要求。

### 需要讨论并修订

1. 购物车与收藏夹 `is_valid` 属性的含义以及必要性。
1. shop::address 是否需要 check
1. 触发器
    需要触发器的场景。（在文档中标明，在 .sql 中实现）
1. 约束
    外键约束能否在各关系模式中指定，而不单独列出？
    当主键的一部分属性是引用外键时，此约束要不要在 .sql 中定义。
1. 索引
    需要讨论并添加
1. user::avatar 使用 `null` 还是 `none.jpg`
1. 冗余
    为保证效率而添加的冗余

### 可能需要修订

1. 各属性 `varchar` 的长度

@billyblondking  @c-rainstorm @ytqqwer

c-rainstorm committed:
Improve: 数据库再修改

1. 确定数据库字符集 utf8mb4
2. 将外键约束添加到建表语句中
3. 修订部分属性最大长度
4. 为部分表添加 `is_valid` 属性
5. 修订触发器，删除不必要的查询. 设计文档中添加触发器描述
6. 为部分表添加索引
7. 重新定义 `user::avatar`
8. **添加商品图片表**

c-rainstorm commented:
最后再检查一下有没有问题，没有问题的话我们今晚把数据库环境搭建起来。

@billyblondking @ytqqwer @rtdarwin

rtdarwin committed:
数据库初稿

rtdarwin committed:
修正目录

rtdarwin committed:
修正逻辑设计物理设计的混淆，同步 .sql 中字符串长度到文档

c-rainstorm committed:
Improve: 不合理的主键设定

1. 为部分表主键添加自增属性
2. 更换部分表的主键为 id
          1. favorite_goods
          2. favorite_shops
          3. shopping_cart
